carapace-server/src/main/java/org/carapaceproxy/server/certificates/DynamicCertificatesManager.java

```diff
@@ -362,9 +391,78 @@ private void certificatesLifecycle() {
+        Order order = acmeClient.createOrderForDomain(domains);
+        cert.setPendingOrderLocation(order.getLocation());
+        LOG.log(Level.INFO, "Pending order location for domains {0}: {1}", new Object[]{domains, order.getLocation()});
+
+        return order;
+    }
+
+    private void createChallengeForCertificateOrder(CertificateData cert, Order order) throws AcmeException {
+        Challenge challenge = acmeClient.getChallengeForOrder(order, cert.isWildcard());
+        if (challenge == null) {
+            cert.setState(VERIFIED);
+        } else {
+            LOG.log(Level.INFO, "Pending challenge data for domain {0}: {1}", new Object[]{cert.getDomain(), challenge.getJSON()});
+            if (cert.isWildcard()) {
+                Dns01Challenge ch = (Dns01Challenge) challenge;
+                r53Client.createDNSChallengeForDomain(cert.getDomain(), ch.getDigest(),
+                        res -> LOG.log(Level.INFO, "Create new TXT DNS challenge-record for domain {0}. Details: {1}", new Object[]{cert.getDomain(), res.responseMetadata()}),
+                        err -> {
+                            dnsChallegeCreationsFailed.add(cert.getDomain());
+                            LOG.log(Level.INFO, "Creation of TXT DNS challenge-record for domain {0} FAILED. Reason: {1}", new Object[]{cert.getDomain(), err});
+                        });
+                cert.setState(DNS_CHALLENGE_WAIT);
+            } else {
+                Http01Challenge ch = (Http01Challenge) challenge;
+                store.saveAcmeChallengeToken(ch.getToken(), ch.getAuthorization()); // used for incoming acme-challenge requests
+                challenge.trigger();
+                cert.setState(VERIFYING);
+            }
+            cert.setPendingChallengeData(challenge.getJSON());
+        }
+    }
+
+    private Challenge getChallengeFromCertificate(CertificateData cert) throws AcmeException {
+        try {
+            JSON challengeData = JSON.parse(cert.getPendingChallengeData());
+            LOG.log(Level.INFO, "CHALLENGE: {0}.", challengeData);
+            return cert.isWildcard()
+                    ? new Dns01Challenge(acmeClient.getLogin(), challengeData)
+                    : new Http01Challenge(acmeClient.getLogin(), challengeData);
+        } catch (IOException | AcmeException e) {
+            return null;
+        }
+    }
+
+    private void checkChallengeResponseForCertificate(Challenge challenge, CertificateData cert) throws AcmeException, IOException {
+        Status status = acmeClient.checkResponseForChallenge(challenge); // checks response and updates the challenge
+        cert.setPendingChallengeData(challenge.getJSON());
+        if (status == Status.VALID) {
+            cert.setState(VERIFIED);
+            cleanupChallengeForCertificate(challenge, cert);
+        } else if (status == Status.INVALID) {
+            cert.setState(REQUEST_FAILED);
+            cleanupChallengeForCertificate(challenge, cert);
+        }
```

Comment:
else ?

Reply:
this has not been addressed

Reply:
there is no else case to manage.. it means that the CA's still processing the validation so we'll check-it out the next time.